nodejs/week4/warmup/index.js

```diff
@@ -0,0 +1,108 @@
+const fs = require('fs');
+
+
+const characterChecker = str => {
+    if (str.length === 0) {
+        return 0
+    }
+    let characterMaker = str.split("");
+    let evaluateLength = str.length;
+    let upperCaseAmount = characterMaker.filter(x => x == x.toUpperCase());
+    let percentageOfUpperCase = upperCaseAmount.length * 100 / evaluateLength;
+    return percentageOfUpperCase
+}
+
+class Email {
+    constructor(subject, body) {
+        this.subject = subject;
+        this.body = body;
+    }
+}
+
+class SpamDetector {
+    constructor(extraWords, extraRules) {
+        this.rules = [
+            {
+                name: "bad word checker",
+                func: this.checkForBadWords
+            },
+            {
+                name: "upper case checker",
+                func: this.evaluateUpperCaseChar
+            }, {
+                name: "subject should not be hello",
+                context: "subject",
+                func: (str) => str.toLowerCase() === "hello"
+            }, {
+                name: "empty string",
+                priority: 0,
+                func: (str) => (!str || str.length === 0)
+
+            }
+        ];
+        this.rules = [...this.rules, ...extraRules.map(f => {
+            return {
+                name: "user defined rules",
+                func: f
+            }
+        })]
+        this.badWords = ["viagra", "free", "offer", "business proposal"];
+        this.badWords = [...this.badWords, ...extraWords];
+    }
+
+    evaluateUpperCaseChar(str) {
+
+        if (characterChecker(str) >= 60) {
+            return true
+        }
+    }
+
+    checkForBadWords(str, self) {
```

Comment:
The same reason will be applied here, Because if I use this inside a callback function it is representing different content (callback function own context)